rust/agama-network/src/dbus/interfaces.rs

```diff
@@ -0,0 +1,402 @@
+        let actions = self.actions.lock();
+        let uuid =
+            Uuid::parse_str(uuid).map_err(|_| NetworkStateError::InvalidUuid(uuid.to_string()))?;
+        actions.send(Action::RemoveConnection(uuid)).unwrap();
+        Ok(())
+    }
+
+    /// Applies the network configuration.
+    ///
+    /// It includes adding, updating and removing connections as needed.
+    pub async fn apply(&self) -> zbus::fdo::Result<()> {
+        let actions = self.actions.lock();
+        actions.send(Action::Apply).unwrap();
+        Ok(())
+    }
+}
+
+/// D-Bus interface for a network connection
+///
+/// It offers an API to query a connection.
+pub struct Connection {
+    connection: Arc<Mutex<NetworkConnection>>,
+}
+
+impl Connection {
+    /// Creates a Connection interface object.
+    ///
+    /// * `UUID`: Connection UUID.
+    pub fn new(connection: Arc<Mutex<NetworkConnection>>) -> Self {
+        Self { connection }
+    }
+
+    /// Returns the underlying connection.
+    fn get_connection(&self) -> MutexGuard<NetworkConnection> {
+        self.connection.lock()
+    }
+}
+
+#[dbus_interface(name = "org.opensuse.Agama.Network1.Connection")]
+impl Connection {
+    #[dbus_interface(property)]
+    pub fn id(&self) -> String {
+        self.get_connection().id().to_string()
+    }
+
+    #[dbus_interface(property, name = "UUID")]
+    pub fn uuid(&self) -> String {
+        self.get_connection().uuid().to_string()
+    }
+}
+
+/// D-Bus interface for IPv4 settings
+pub struct Ipv4 {
+    actions: Arc<Mutex<Sender<Action>>>,
+    connection: Arc<Mutex<NetworkConnection>>,
+}
+
+impl Ipv4 {
+    /// Creates an IPv4 interface object.
+    ///
+    /// * `actions`: sending-half of a channel to send actions.
+    /// * `connection`: connection to expose over D-Bus.
+    pub fn new(actions: Sender<Action>, connection: Arc<Mutex<NetworkConnection>>) -> Self {
+        Self {
+            actions: Arc::new(Mutex::new(actions)),
+            connection,
+        }
+    }
+
+    /// Returns the underlying connection.
+    fn get_connection(&self) -> MutexGuard<NetworkConnection> {
+        self.connection.lock()
+    }
+
+    /// Updates the connection data in the NetworkSystem.
+    ///
+    /// * `connection`: Updated connection.
+    fn update_connection(
+        &self,
+        connection: MutexGuard<NetworkConnection>,
+    ) -> zbus::fdo::Result<()> {
+        let actions = self.actions.lock();
+        actions
+            .send(Action::UpdateConnection(connection.clone()))
+            .unwrap();
+        Ok(())
+    }
+}
+
+#[dbus_interface(name = "org.opensuse.Agama.Network1.Connection.IPv4")]
+impl Ipv4 {
+    #[dbus_interface(property)]
+    pub fn addresses(&self) -> Vec<(String, u32)> {
+        let connection = self.get_connection();
+        connection
+            .ipv4()
+            .addresses
+            .iter()
+            .map(|(addr, prefix)| (addr.to_string(), *prefix))
```

Comment:
so no netmask support, right?

Reply:
You use an address and a prefix ("192.168.1.0/32"), so no need for a network mask.